File: TopInterview150/42_trapping_water/trap.go
func trap(height []int) int {
	res, n := 0, len(height)-1
	l, r, lMax, rMax := 0, n, 0, height[n]

	for l <= r {
		lMax = max(height[l], lMax)
		rMax = max(height[r], rMax)

		if rMax >= lMax {
			res += lMax - height[l]
			l++
		} else {
			res += rMax - height[r]
			r--
		}
	}

	return res
}

func max(a int, b int) int {
	if a > b {
		return a
	} else {
		return b
	}
}

File: TopInterview150/42_trapping_water/trap.java
class Solution {
    public int trap(int[] height) {
        int result = 0;

        int n = height.length;
        int left = 0;
        int right = n - 1;

        int leftMax = height[0];
        int rightMax = height[n - 1];

        while (left <= right) {
            leftMax = Math.max(height[left], leftMax);
            rightMax = Math.max(height[right], rightMax);

            if (rightMax >= leftMax) {
                result += leftMax - height[left];
                left++;
            } else {
                result += rightMax - height[right];
                right--;
            }
        }

        return result;
    }
}